Predictive Maintenance of Equipment

siddhant4ds opened the pull request:
# Pull Request for ML-Crate 💡


## Issue Title: <!-- Enter the issue title here --> Predictive Maintenance Equipment Instruments

- **Info about the related issue (Aim of the project)** : <!-- What's the goal of the project --> Detecting machine failure based on different operational parameters for predictive maintenance of equipment. The dataset is modeled after an existing milling machine and consists of 10,000 data points stored as rows with 14 features in columns. Only 339 out of 10000 data points represent machine failure, making it an extremely imbalanced classification problem.

- **Name:** <!--Mention Your name--> Siddhant Tiwari
- **Email ID for further communication**: <!--Mention your email ID--> siddhant.tiwari.ds@gmail.com
- **GitHub ID:** <!-- Mention your GitHub ID --> siddhant4ds
- **Identify yourself:** <!-- Mention your role --> SSOC-3 (2024) Contributor


<!-- Mention the following details and these are mandatory -->

Closes: #633 
 
### Describe the add-ons or changes you've made 📃

1. Exploratory analysis of features along with preprocessing.
2. Feature engineering based on domain knowledge of different types of machine failures.
3. Feature selection based on statistical tests.
4. Created a holdout set for testing all the models using Stratified sampling to maintain imbalance ratio.
5. Training and validation of multiple types and configurations of: linear models, support vector machines, tree models, tree ensembles, gradient-boosting trees and neural networks.
6. Models were tuned and evaluated based on F2-score, which is a harmonic, weighted mean of precision and recall, with recall having twice the importance of precision. This is because identifying all instances of machine failure correctly is critical to us.
7. Created a Streamlit web app to provide an interface for testing new data points with appropriate input methods. For inference, the best model is chosen i.e., AdaBoostClassifier trained on full feature set.  

(Also renamed the project due to incorrect issue title)

## Type of change ☑️

What sort of change have you made:
<!--
Example how to mark a checkbox:-
- [x] My code follows the code style of this project.
-->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, local variables)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested? ⚙️

Reproducibility: multiple runs of machine learning models to produce same results.
Dependencies: running on local environment as well as cloud-hosted notebooks (Kaggle & Colab)
Web app: user-testing

## Checklist: ☑️
<!--
Example how to mark a checkbox:-
- [x] My code follows the code style of this project.
-->
- [x] My code follows the guidelines of this project.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly wherever it was hard to understand.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] I have added things that prove my fix is effective or that my feature works.
- [x] Any dependent changes have been merged and published in downstream modules.

abhisheks008 commented:
![image](https://github.com/abhisheks008/ML-Crate/assets/68724349/6c945365-dd59-4ff4-a6a4-ba7fc4317eb6)

Please add the demo video inside the README and check the images are not visible. Just copy the video and paste it inside the README, don't try to hyperlink the video.

@siddhant4ds

siddhant4ds commented:
@abhisheks008 there was some issue with the relative path for images. I have fixed that and they should be visible. Also added the video using the method you suggested. Let me know if any other changes are required. Thanks

abhisheks008 reviewed (approved):
Approved @siddhant4ds
